File: algo/threeNumberSum.py
"""
https://www.algoexpert.io/questions/Three%20Number%20Sum
"""
def threeNumberSum(array, targetSum):
    # Write your code here.
     ans = []
     for num1 in range(len(array)):
          for num2 in range(len(array)):
               if num1 != num2:
                    num3 = targetSum - array[num1] - array[num2]
                    if num3 in array and (array[num2] != num3 and array[num1] != num3):
                         ans.append(sorted([array[num1], array[num2], num3]))
     r = []
     for sublist in ans:
          if sublist not in r:
               r.append(sublist)
     print(r)


array = [12,3,1,2,-6,5,-8,6]
target = 0
threeNumberSum(array, target)

File: algo/threeNumberSum.js
function threeNumberSum(array, targetSum){
    //sort the array
    array.sort((a,b) => a - b);
    const ans = []
    for(let i = 0; i < array.length; i++){
        let left = i + 1;
        let right = array.length - 1
        while(left < right){
            sumTotal = array[i] + array[right] + array[left]
            if (sumTotal == target){
                ans.push([array[i], array[right], array[left]])
                left ++;
            }
            else if (sumTotal < targetSum){
                left ++;
            }else if (sumTotal > targetSum){
                right -- ;
            }
        }
    }
    return ans
}

target = 0
array = [12,3,1,2,-6,5,-8,6]

console.log(threeNumberSum(array, target))